src/lib.rs

```diff
@@ -77,15 +90,30 @@ use tokio::sync::{oneshot, OwnedMutexGuard};
 /// * `Key` batch key type.
 /// * `Item` single item type to be batched together into a `Vec<Item>`.
 /// * `T` value returned by [`BatchResult::Done`], default `()`.
-#[derive(Clone)]
-pub struct BatchMutex<Key: Eq + Hash, Item, T = ()> {
-    queue: Arc<DashMap<Key, BatchState<Key, Item, T>>>,
+pub struct BatchMutex<Key, Item, T = ()> {
+    hasher: RandomState,
+    queue: Arc<[Shard<Key, Item, T>]>,
+}
+type Shard<K, I, T> = Mutex<HashMap<K, BatchState<I, T>>>;
+
+impl<Key, Item, T> Clone for BatchMutex<Key, Item, T> {
+    fn clone(&self) -> Self {
+        Self {
+            hasher: self.hasher.clone(),
+            queue: self.queue.clone(),
+        }
+    }
 }
 
 impl<Key: Eq + Hash, Item, T> Default for BatchMutex<Key, Item, T> {
     fn default() -> Self {
+        let cap =
+            (std::thread::available_parallelism().map_or(1, usize::from) * 4).next_power_of_two();
+        let mut shards = Vec::with_capacity(cap);
+        shards.resize_with(cap, || Mutex::new(HashMap::new()));
         Self {
-            queue: <_>::default(),
+            hasher: RandomState::new(),
+            queue: shards.into_boxed_slice().into(),
         }
```

Comment:
This creates a more optimised version of dashmap for our needs.

2 optimisations:
1. Uses a `Mutex` instead of a `RwLock` since we never use read-only. `Mutex` is faster in this case
2. `Arc<DashMap>` has an internal `Box<[Shards]>` allocation. Moving it up to `Arc<[Shards]>` removes a pointer indirection

Reply:
Hmm I'm not sure I want to implement dashmap sharding here though, I replaced my original `Mutex<Map` impl with `DashMap` which gives more optimal performance without extra (local) code complexity.

If something like `DashMap` didn't exist I'd probably just keep using `Mutex<Map`.

Is there a good reason to want to avoid dashmap?